File: tools/internal/pkg/grouping/grouping_test.go
//
// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
//
// See LICENSE.txt for license information
//

package grouping

import (
	"strings"
	"testing"
)

func TestSplitting(t *testing.T) {
	tests := []struct {
		elts       []int
		splitIndex int
		result     [][]int
	}{

		{
			elts:       []int{0, 1, 2},
			splitIndex: 1,
			result: [][]int{
				[]int{0},
				[]int{1, 2},
			},
		},
		{
			elts:       []int{1, 2, 3},
			splitIndex: 2,
			result: [][]int{
				[]int{1, 2},
				[]int{3},
			},
		},
	}

	for _, tt := range tests {
		e := Init()
		for j := 0; j < len(tt.elts); j++ {
			err := e.AddDatapoint(j, tt.elts)
			if err != nil {
				t.Fatalf("unable to add point: %s", err)
			}
		}

		t.Logf("Group ready, splitting at %d", tt.splitIndex)
		ng, err := e.splitGroup(e.Groups[0], tt.splitIndex, tt.elts)
		if err != nil {
			t.Fatalf("unable to split group: %s", err)
		}
		if len(ng.Elts) != len(tt.result[1]) {
			str := groupToString(ng.Elts)
			t.Fatalf("result group is of length %d instead of the expected %d (%s)", len(ng.Elts), len(tt.result[1]), str)
		}

		// We check the content of the first group
		if len(e.Groups[0].Elts) != len(tt.result[0]) {
			t.Fatalf("invalid first group after split: %d elements vs. %d expected", len(e.Groups[0].Elts), len(tt.result[0]))
		}

		for g := 0; g < 2; g++ {
			for n := 0; n < len(e.Groups[g].Elts); n++ {
				if tt.elts[e.Groups[g].Elts[n]] != tt.result[g][n] {
					t.Fatalf("Element %d of result group #%d is %d instead of expected %d (group is:%s)", n, g, e.Groups[g].Elts[n], tt.result[g][n], groupToString(e.Groups[g].Elts))
				}
			}
		}
	}
}

func TestGrouping(t *testing.T) {
	tests := []struct {
		points       []int
		groupsResult [][]int
	}{
		{
			points: []int{1, 2, 3, 3, 3},
			groupsResult: [][]int{
				[]int{1, 2},
				[]int{3, 3, 3},
			},
		},
		{
			points: []int{1, 2, 3},
			groupsResult: [][]int{
				[]int{1, 2, 3},
			},
		},
		{
			points: []int{1, 2, 3, 5},
			groupsResult: [][]int{
				[]int{1, 2, 3, 5},
			},
		},
		{
			points: []int{1, 2, 3, 10, 11, 12},
			groupsResult: [][]int{
				[]int{1, 2, 3},
				[]int{10, 11, 12},
			},
		},
		{
			points: []int{0, 1, 2, 5, 6, 7, 20, 30, 25},
			groupsResult: [][]int{
				[]int{0, 1, 2},
				[]int{5, 6, 7},
				[]int{20, 25, 30},
			},
		},
		{
			points: []int{100, 0, 1, 5, 6, 7, 20, 2, 30, 25},
			groupsResult: [][]int{
				[]int{0, 1, 2},
				[]int{5, 6, 7},
				[]int{20, 25, 30},
				[]int{100},
			},
		},
	}

	num := 1
	for _, tt := range tests {
		e := Init()
		t.Logf("Running test %d", num)
		for j := 0; j < len(tt.points); j++ {
			t.Logf("-> Adding %d\n", tt.points[j])
			err := e.AddDatapoint(j, tt.points)
			if err != nil {
				t.Fatalf("unable to add point: %s", err)
			}
		}

		// Compare the resulting groups with what we expect
		gps, err := e.GetGroups()
		if err != nil {
			t.Logf("unable to get groups: %s", err)
		}

		var groupsStr []string
		for _, g := range gps {
			groupsStr = append(groupsStr, groupToString(g.Elts))
		}
		if len(gps) != len(tt.groupsResult) {
			t.Fatalf("test %d reports %d groups instead of %d (groups: %s)\n", num, len(gps), len(tt.groupsResult), strings.Join(groupsStr, "; "))
		}
		for k := 0; k < len(tt.groupsResult); k++ {
			for l := 0; l < len(gps[k].Elts); l++ {
				if len(gps[k].Elts) != len(tt.groupsResult[k]) {
					t.Fatalf("returned group #%d has %d elements while expecting %d\n", k, len(gps[k].Elts), len(tt.groupsResult[k]))
				}
				if tt.groupsResult[k][l] != tt.points[gps[k].Elts[l]] {
					t.Fatalf("element %d of group %d is %d instead of %d\n", l, k, tt.points[gps[k].Elts[l]], tt.groupsResult[k][l])
				}
			}
		}
		num++
	}
}

File: src/common/grouping_test.c
/*************************************************************************
 * Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
 *
 * See LICENSE.txt for license information
 ************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "grouping.h"

#define MAX_SUBGROUPS (4)
#define MAX_ELTS (10)

typedef struct gp_result
{
    int size;
    int elts[MAX_ELTS];
} gp_result_t;

typedef struct gp_test
{
    int num_points;
    int points[MAX_ELTS];
    int num_gps;
    gp_result_t groups_result[MAX_SUBGROUPS];
} gp_test_t;

static int grouping_test(void)
{
    gp_test_t tests[] = {
        {
            num_points : 5,
            points : {1, 2, 3, 3, 3},
            num_gps : 2,
            groups_result : {
                {
                    size : 2,
                    elts : {1, 2},
                },
                {
                    size : 3,
                    elts : {3, 3, 3},
                },
            },
        },
        {
            num_points : 3,
            points : {1, 2, 3},
            num_gps : 1,
            groups_result : {
                {
                    size : 3,
                    elts : {1, 2, 3},
                },
            },
        },
        {
            num_points : 4,
            points : {1, 2, 3, 5},
            num_gps : 1,
            groups_result : {
                {
                    size : 4,
                    elts : {1, 2, 3, 5},
                },
            },
        },
        {
            num_points : 6,
            points : {1, 2, 3, 10, 11, 12},
            num_gps : 2,
            groups_result : {
                {
                    size : 3,
                    elts : {1, 2, 3},
                },
                {
                    size : 3,
                    elts : {10, 11, 12},
                },
            },
        },
        {
            num_points : 9,
            points : {0, 1, 2, 5, 6, 7, 20, 30, 25},
            num_gps : 3,
            groups_result : {
                {
                    size : 3,
                    elts : {0, 1, 2},
                },
                {
                    size : 3,
                    elts : {5, 6, 7},
                },
                {
                    size : 3,
                    elts : {20, 25, 30},
                },
            },
        },
        {
            num_points : 10,
            points : {100, 0, 1, 5, 6, 7, 20, 2, 30, 25},
            num_gps : 4,
            groups_result : {
                {
                    size : 3,
                    elts : {0, 1, 2},
                },
                {
                    size : 3,
                    elts : {5, 6, 7},
                },
                {
                    size : 3,
                    elts : {20, 25, 30},
                },
                {
                    size : 1,
                    elts : {100},
                },
            },
        },
    };

    int i, j, k, l;
    for (i = 0; i < 6; i++)
    {
        grouping_engine_t *e;
        if (grouping_init(&e))
        {
            fprintf(stderr, "[ERROR] unable to initialize grouping\n");
            return 1;
        }
        fprintf(stdout, "*** Running test %d\n", i);
        for (j = 0; j < tests[i].num_points; j++)
        {
            fprintf(stdout, "-> Adding %d\n", tests[i].points[j]);
            add_datapoint(e, j, tests[i].points);
        }

        // Compare the resulting groups with what we expect
        group_t *gps = NULL;
        int num_gps = 0;
        get_groups(e, &gps, &num_gps);
        if (num_gps != tests[i].num_gps)
        {
            fprintf(stderr, "*** [ERROR] Test %d reports %d groups instead of %d\n", i, num_gps, tests[i].num_gps);
            return 1;
        }
        group_t *ptr = gps;
        for (k = 0; k < num_gps; k++)
        {
            for (l = 0; l < ptr->size; l++)
            {
                if (ptr->size != tests[i].groups_result[k].size)
                {
                    fprintf(stderr, "*** [ERROR] Returned group #%d has %d elements while expecting %d\n", k, ptr->size, tests[i].groups_result[k].size);
                    return 1;
                }
                if (tests[i].points[ptr->elts[l]] != tests[i].groups_result[k].elts[l])
                {
                    fprintf(stderr, "*** [ERROR] element %d of group %d is %d instead of %d\n", l, k, ptr->elts[l], tests[i].groups_result[k].elts[l]);
                    return 1;
                }
            }
            ptr = ptr->next;
        }
        grouping_fini(&e);
        fprintf(stdout, "*** Test %d successful\n", i);
    }

    return 0;
}

int main(int argc, char **argv)
{
    if (grouping_test())
    {
        fprintf(stderr, "[ERROR] grouping test failed\n");
        return EXIT_FAILURE;
    }
    fprintf(stdout, "grouping test succeeded\n");

    return EXIT_SUCCESS;
}
